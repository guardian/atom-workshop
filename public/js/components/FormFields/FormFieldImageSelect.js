import React, { PropTypes } from 'react';
import filesize from 'filesize';
import Modal from '../Utilities/Modal';
import ShowErrors from '../Utilities/ShowErrors';
import { errorPropType } from '../../constants/errorPropType';
import {parseImageFromGridCrop, findSmallestAssetAboveWidth, gridUrlFromApiUrl} from '../../util/imageHelpers';
import {logInfo} from '../../util/logger';

const assetPropType = PropTypes.shape({
  mimeType: PropTypes.string.isRequired,
  file: PropTypes.string.isRequired,
  dimensions: PropTypes.shape({
    height: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired
  }).isRequired,
  size: PropTypes.number.isRequired
});

const gridImagePropType = PropTypes.shape({
  mediaId: PropTypes.string.isRequired,
  assets: PropTypes.arrayOf(assetPropType).isRequired,
  master: assetPropType.isRequired
});

class FormFieldImageSelect extends React.Component {

  static propTypes = {
    onUpdateField: PropTypes.func,
    fieldValue: gridImagePropType,
    fieldName: PropTypes.string,
    fieldLabel: PropTypes.string,
    fieldErrors: PropTypes.arrayOf(errorPropType),
    formRowClass: PropTypes.string,
    gridUrl: PropTypes.string.isRequired
  }

  state = {
    modalOpen: false
  }

  closeModal = () => {
    this.setState({ modalOpen: false });
    window.removeEventListener('message', this.onMessage, false);
  }

  openModal = () => {
      this.setState({ modalOpen: true });
      window.addEventListener('message', this.onMessage, false);
  }

  validMessage(data) {
    return data && data.crop && data.crop.data && data.image && data.image.data;
  }

  onMessage = (event) => {

      if (event.origin !== this.props.gridUrl) {
          logInfo("didn't come from the grid");
          return;
      }

      const data = event.data;

      if (!data) {
          logInfo("got no data...");
          return;
      }

      if (!this.validMessage(data)) {
          logInfo("not a valid message...");
          return;
      }

      this.closeModal();
      this.props.onUpdateField(parseImageFromGridCrop(data.crop.data));
  }

  renderWithoutImage() {
    return (
      <button className="image-select__button" onClick={this.openModal}>
          + Add Image from Grid
      </button>
    );
  }

  renderWithImage() {
    const thumbnail = findSmallestAssetAboveWidth(this.props.fieldValue.assets, 200);
    return (
      <div className="image-select__image-display">
        <div className="image-select__image">
          <img src={thumbnail.file} />
        </div>
        <div className="image-select__image-details">
          <div className="image-select__image-details__detail">
            <a target="_blank" href={gridUrlFromApiUrl(this.props.fieldValue.mediaId)}>View image in the Grid</a>
          </div>
          <div className="image-select__image-details__detail">
            Size: {filesize(this.props.fieldValue.master.size)}
          </div>
          <div className="image-select__image-details__detail">
            Dimensions: {this.props.fieldValue.master.dimensions.width} x {this.props.fieldValue.master.dimensions.height}
          </div>
          <button className="image-select__button" type="button" onClick={this.openModal}>
              Replace Image
          </button>
        </div>
      </div>
    );
  }

  render () {
    return (
      <div className={this.props.formRowClass || "form__row"}>
        {this.props.fieldLabel ? <label htmlFor={this.props.fieldName} className="form__label">{this.props.fieldLabel}</label> : false}
        <div className="image-select">
            {this.props.fieldValue ? this.renderWithImage() : this.renderWithoutImage()}

            <Modal isOpen={this.state.modalOpen} dismiss={this.closeModal}>
                <iframe className="image-select__modal" src={this.props.gridUrl}></iframe>
            </Modal>
        </div>
        <ShowErrors errors={this.props.fieldErrors}/>
      </div>
    );
  }
}

export default FormFieldImageSelect;
