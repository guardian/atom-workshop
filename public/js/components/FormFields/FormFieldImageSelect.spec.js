import React from 'react';
import FormFieldImageSelect from './FormFieldImageSelect';
import renderer from 'react-test-renderer';
import {configure, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';

configure({ adapter: new Adapter() });

const gridUrl = "https://media.test.dev-gutools.co.uk";

const testImage = {
  "assets":[
    {
      "file":"https://s3-eu-west-1.amazonaws.com/media-origin.test.dev-guim.co.uk/5dfc898f6866c8bb4224b65b9aae89e6eb7af30f/783_0_2025_1215/2000.jpg",
      "mimeType":"image/jpeg",
      "size":119846,
      "dimensions":{"width":2000,"height":1200}
    },
    {
      "file":"https://s3-eu-west-1.amazonaws.com/media-origin.test.dev-guim.co.uk/5dfc898f6866c8bb4224b65b9aae89e6eb7af30f/783_0_2025_1215/1000.jpg",
      "mimeType":"image/jpeg",
      "size":35608,
      "dimensions":{"width":1000,"height":600}
    }
  ],
  "master":{
    "file":"https://s3-eu-west-1.amazonaws.com/media-origin.test.dev-guim.co.uk/5dfc898f6866c8bb4224b65b9aae89e6eb7af30f/783_0_2025_1215/master/2025.jpg",
    "mimeType":"image/jpeg",
    "size":980347,
    "dimensions":{"width":2025,"height":1215}
  },
  "mediaId":"https://api.media.test.dev-gutools.co.uk/images/5dfc898f6866c8bb4224b65b9aae89e6eb7af30f"
};

const mockGridMessage = {"image":{"uri":"https://api.media.test.dev-gutools.co.uk/images/8582a3fb775d38294d61a35fa036338518d8e375","data":{"id":"8582a3fb775d38294d61a35fa036338518d8e375","uploadTime":"2017-02-03T11:37:53Z","uploadedBy":"pa","lastModified":"2017-02-03T11:37:53Z","identifiers":{},"uploadInfo":{"filename":"POLITICS_Scotland_113464.jpg"},"source":{"file":"http://media-service-test-imagebucket-1qt2lbcwnpgl0.s3.amazonaws.com/8/5/8/2/a/3/8582a3fb775d38294d61a35fa036338518d8e375","size":574944,"mimeType":"image/jpeg","dimensions":{"width":2948,"height":2161},"secureUrl":"https://media-service-test-imagebucket-1qt2lbcwnpgl0.s3.amazonaws.com/8/5/8…UHJNQ6MS2A&Expires=1486123800&Signature=VYh2qy5lX%2BQYd1Gh%2Bx1eBpDhgnU%3D"},"thumbnail":{"file":"http://media-service-test-thumbbucket-1sm53rniszbqf.s3.amazonaws.com/8/5/8/2/a/3/8582a3fb775d38294d61a35fa036338518d8e375","size":10761,"mimeType":"image/jpeg","dimensions":{"width":256,"height":188},"secureUrl":"https://d2av06b16cc3yv.cloudfront.net/8/5/8/2/a/3/8582a3fb775d38294d61a35fa…AdRoPIMkIxRAS1J2SOhFBwr7m1ChKRKbZw4Y34g__&Key-Pair-Id=APKAJPTTPZNNPHQSSUAQ"},"fileMetadata":{},"metadata":{"dateTaken":"2017-02-03T00:00:00Z","description":"File photo dated 07/07/16 of Michael Gove, who has said that Nicola Sturgeon would be \"foolish\" to call a second independence referendum in the wake of the Brexit vote. PRESS ASSOCIATION Photo. Issue date: Friday February 3, 2017. Gove, who served as both education and justice secretary under David Cameron, said: \"The people of the United Kingdom, having voted to leave one union that didn't work, the people of Scotland are not going to vote to leave another union that works.\" See PA story POLITICS Scotland. Photo credit should read: Jonathan Brady/PA Wire","credit":"PA","byline":"Jonathan Brady","title":"Brexit","copyright":"PA Archive/PA Images","suppliersReference":"POLITICS_Scotland_113464.JPG","source":"PA","specialInstructions":"FILE PHOTO","keywords":["politics conservatives"],"city":"London","country":"United Kingdom"},"originalMetadata":{"dateTaken":"2017-02-03T00:00:00Z","description":"File photo dated 07/07/16 of Michael Gove, who has said that Nicola Sturgeon would be \"foolish\" to call a second independence referendum in the wake of the Brexit vote. PRESS ASSOCIATION Photo. Issue date: Friday February 3, 2017. Gove, who served as both education and justice secretary under David Cameron, said: \"The people of the United Kingdom, having voted to leave one union that didn't work, the people of Scotland are not going to vote to leave another union that works.\" See PA story POLITICS Scotland. Photo credit should read: Jonathan Brady/PA Wire","credit":"PA","byline":"Jonathan Brady","title":"Brexit","copyright":"PA Archive/PA Images","suppliersReference":"POLITICS_Scotland_113464.JPG","source":"PA","specialInstructions":"FILE PHOTO","keywords":["politics conservatives"],"city":"London","country":"United Kingdom"},"usageRights":{"category":"agency","supplier":"PA"},"originalUsageRights":{"category":"agency","supplier":"PA"},"exports":[],"usages":{},"leases":{},"collections":[],"userMetadata":{},"valid":true,"invalidReasons":{},"cost":"free","persisted":{"value":false,"reasons":[]}}},"crop":{"uri":"https://cropper.media.test.dev-gutools.co.uk/crops","data":{"id":"1194_688_1072_643","author":"christopher.lloyd@guardian.co.uk","date":"2017-02-03T11:53:01Z","specification":{"uri":"https://api.media.test.dev-gutools.co.uk/images/8582a3fb775d38294d61a35fa036338518d8e375","bounds":{"x":1194,"y":688,"width":1072,"height":643},"aspectRatio":"5:3","type":"crop"},"master":{"file":"http://media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/master/1072.jpg","size":682941,"mimeType":"image/jpeg","dimensions":{"width":1072,"height":643},"secureUrl":"https://s3-eu-west-1.amazonaws.com/media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/master/1072.jpg"},"assets":[{"file":"http://media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/1000.jpg","size":77716,"mimeType":"jpg","dimensions":{"width":1000,"height":600},"secureUrl":"https://s3-eu-west-1.amazonaws.com/media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/1000.jpg"},{"file":"http://media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/500.jpg","size":25313,"mimeType":"jpg","dimensions":{"width":500,"height":300},"secureUrl":"https://s3-eu-west-1.amazonaws.com/media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/500.jpg"},{"file":"http://media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/140.jpg","size":6146,"mimeType":"jpg","dimensions":{"width":140,"height":84},"secureUrl":"https://s3-eu-west-1.amazonaws.com/media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/140.jpg"},{"file":"http://media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/1072.jpg","size":95745,"mimeType":"jpg","dimensions":{"width":1072,"height":643},"secureUrl":"https://s3-eu-west-1.amazonaws.com/media-origin.test.dev-guim.co.uk/8582a3fb775d38294d61a35fa036338518d8e375/1194_688_1072_643/1072.jpg"}]}}};

test('Should render no image when none provided', () => {
  const component = renderer.create(
    <FormFieldImageSelect
      gridUrl={gridUrl}
      fieldName="test field name"
      fieldLabel="test field label"
      onUpdateField={() => {}}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Should render with image when provided', () => {
  const component = renderer.create(
    <FormFieldImageSelect
      gridUrl={gridUrl}
      fieldName="test field name"
      fieldLabel="test field label"
      fieldValue={testImage}
      onUpdateField={() => {}}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Should save image when message recieved', () => {

  const fn = jest.fn();

  const imageSelect = mount(
    <FormFieldImageSelect
      gridUrl={gridUrl}
      fieldName="test field name"
      fieldLabel="test field label"
      onUpdateField={fn}
    />
  );

  imageSelect.instance().onMessage({
    data: mockGridMessage,
    origin: gridUrl
  });

  expect(fn).toHaveBeenCalled();

});

test('Should not save image when message origin is inCorrect', () => {

  const fn = jest.fn();

  const imageSelect = mount(
    <FormFieldImageSelect
      gridUrl={gridUrl}
      fieldName="test field name"
      fieldLabel="test field label"
      onUpdateField={fn}
    />
  );

  imageSelect.instance().onMessage({
    data: mockGridMessage,
    origin: "not the gridUrl"
  });

  expect(fn).not.toHaveBeenCalled();

});
