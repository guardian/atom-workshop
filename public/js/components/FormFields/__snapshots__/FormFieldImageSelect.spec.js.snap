exports[`test Should render no image when none provided 1`] = `
<div>
  <label
    className="form__label">
    test field label
  </label>
  <div
    className="image-select form__group">
    <button
      className="image-select__button"
      onClick={[Function]}>
      + Add Image from Grid
    </button>
  </div>
  <div />
</div>
`;

exports[`test Should render with image when provided 1`] = `
<div>
  <label
    className="form__label">
    test field label
  </label>
  <div
    className="image-select form__group">
    <div
      className="image-select__image-display">
      <div
        className="image-select__image">
        <img
          src="https://s3-eu-west-1.amazonaws.com/media-origin.test.dev-guim.co.uk/5dfc898f6866c8bb4224b65b9aae89e6eb7af30f/783_0_2025_1215/1000.jpg" />
      </div>
      <div
        className="image-select__image-details">
        <div
          className="image-select__image-details__detail">
          <a
            href="https://media.test.dev-gutools.co.uk/images/5dfc898f6866c8bb4224b65b9aae89e6eb7af30f"
            target="_blank">
            View image in the Grid
          </a>
        </div>
        <div
          className="image-select__image-details__detail">
          Size: 
          957.37 KB
        </div>
        <div
          className="image-select__image-details__detail">
          Dimensions: 
          2025
           x 
          1215
        </div>
        <button
          className="image-select__button"
          onClick={[Function]}>
          Replace Image
        </button>
      </div>
    </div>
  </div>
  <div />
</div>
`;
